web/src/engine/predictive-text/worker-thread/src/main/correction/context-tokenization.ts

```diff
@@ -430,8 +435,17 @@ export class ContextTokenization {
     // accordingly.
     const tailIndex = 0 - (stackedDeletes.length - 1) + (editBoundary.omitsEmptyToken ? -1 : 0);
     // Mutates stackedInserts, stackedDeletes.
+    const baseRemovedTokenCount = Math.max(0, stackedDeletes.length - stackedInserts.length);
     const transformMap = assembleTransforms(stackedInserts, stackedDeletes, tailIndex);
 
+    // If there's an empty transform in the 0 position and we already know we're
+    // dropping tokens - and only deleting - we're dropping an
+    // otherwise-untracked empty token - make sure it's included!
+    const droppedFinalTransform = baseRemovedTokenCount > 0 && transform.insert == '' && TransformUtils.isEmpty(transformMap.get(0));
+    // Past that, if we have more delete entries than insert entries for our transforms, we
+    // dropped some tokens outright.
+    const removedTokenCount = baseRemovedTokenCount + (droppedFinalTransform ? 1 : 0);
```

Comment:
Dropping an empty token at the tail has the same transform as doing nothing to that token, so the logic's a bit involved here.  In particular - if "base removed" count is 0, then we couldn't have dropped an empty token at the tail.  If it's greater than 0, we had to have dropped the tail to get to other removed tokens.